### ISSUE TITLE
IDC: Needs more graceful handing of mapping domains

### ISSUE THREAD

**Issue by kraftbj** [Automattic/jetpack#2702]
When using Donncha's Domain Mapping plugin (possibly others), if the option to use the "original" URL for admin is checked, then IDC may trip all of the time (since we may have the site at the mapped domain and the site itself is reporting the "original" domain).

Since upon connection, it would have used the "original" name, this is likely a corner case of folks who setup Jetpack on the mapped domain, then later enabled the option to force the original domain for admin.

Open question: How do we want to address this type of case? Kill the local check if the primary mapped domain != the siteurl (local) on multisite's with domain mapping enabled?

cc: @dereksmart

**Comment by richardmtl** [comment]
This came up in 2342326-t

**Comment by RavanH** [comment]
Hi guys, thanks for picking this up!

Some thoughts: 

1. There are many possibilities a multisite with Domain Mapping can be set up, and depending on the domain mapping plugin used, there may even be fundamental differences between sites on the same network. A site may have both back and front end on stricktly one domain, or it may have front end on the mapped domain while back end on another (most likely the original network domain but can be something else too!) or the site may have multiple domains mapped and each domain may be used independently (without redirect) to access the site (front and/or back end). To complicate things even more: It may not even a domain mapping plugin but also a multilingual plugin like Polylang which allows different languages be served over different domains.

2. Current Jetpack may actually be working as close as possible to any desired situation, but the current message and the choices it presents to resolve the (sudden or not) discrepancy between perceived 'old' and 'new' domain, may just need some clarification. What does Jetpack see as the old and the new situation and what are the consequences of each possible resolution?  Most importantly: how does a site admin resolve the issue ( or dismiss it? ) without loosing current state and subscribers...

Hope this does not make things even more confusing :)

**Comment by chaselivingston** [comment]
Also reported in 2345839-t

**Comment by kraftbj** [comment]
Thanks @RavanH. Right now, Jetpack is comparing the siteurl and homeurl option value on the site and what we have saved up on wp.com when prompting the admin. Some of this may be older changes that, now that we're alerting folks to something being unexpected, something that changed long ago is mucking things up a bit.

I haven't been able to dedicate a lot of time to testing the various possibilities specific to domain mapping options yet but I am curious to see how the various plugins impact the site/home url and how it is reported.

**Comment by richardmtl** [comment]
another in 2342106-t

**Comment by kraftbj** [comment]
HEs: Can we add as much detail as possible on these, instead of just the ticket link? I'd like to get a sense of how multisite domain mapping is being implemented in each, etc.

For 2342106-t: Donncha's plugin, options Permanent redirect, User domain mapping page, Redirect administration pages to site's original domain.

Generally, JP will use the home/site URL that is provided by the site. Since we've never prompted for this before, we may get a lot of old stuff that, if they connected today, they would have never seen.

**Comment by richardmtl** [comment]
For 2342106-t:

```https://wordpress.org/plugins/wordpress-mu-domain-mapping/

As far as configuration we have our server IP Addresses in the network
configuration page along with these options activated...

1. Permanent redirect (better for your blogger's pagerank)
2. User domain mapping page
3. Redirect administration pages to site's original domain (remote
login disabled if this redirect is disabled)

On the "Domains" tab we have two domains per site id, the www and non-www
subdomains.

For instance, site ID #5 has domain.com as a primary domain and
www.domain.com as a non-primary domain.```

**Comment by kraftbj** [comment]
Discussed with @dereksmart, @richardmtl in Slack. For 3.7.1, let's disable the IDC check when `defined( 'SUNRISE' )` which will generally exclude all domain mapping plugins/installs.

It is generating higher-than-it-helps support load for these cases and some additional thought need to go into the best way to prompt when there is an issue and not when there isn't.

**Comment by RavanH** [comment]
@kraftbj while I understand that choice I fear that may lead to an indefinite postponement. There may be valid cases where a site switches domain name and Jetpack's data needs updating accordingly. I still have the feeling that it's only the message that needs a bit more clarity and maybe a "leave as it is / ignore" option... And maybe an option on the Debug page to get back to that message at a later stage.

**Comment by kraftbj** [comment]
@RavanH It isn't indefinite postponement. We're getting a decent amount of support traffic for sites with domain mapping plugins that there isn't a clear answer as to 1. why this is happening, 2. what is the expected result, etc.

Switching it off while researching the issue returns the status quo from the beginning of Jetpack. As an an alternative, we could switch it off for stable versions of Jetpack using domain mapping while keeping it enabled for development versions so folks techy enough to follow development, use betas, etc can continue to easily help us figure out the rough edges without impacting the more regular users who are writing in.

If there is a legit issue, leaving it as is/ignore is not an option, long term, as the problem will resurface at some point in a production-impacting way (hence why we added this in the first place... folks running a cloned site with the same IDs on two sites for years cycles their connection, loses old stats, and only then realize that their stats for both sites have been combined without the ability to separate for years, etc).

With domain mapping, there are multiple URLs that are both completely legit for a particular site, but the open question on which one should Jetpack use? How does the Domain Mapping plugins work with regard to reporting their URLs to Jetpack? What could change that? Are there more common plugins besides the three I know about offhand? Some of the edge cases are sites that on virtually every admin load are trying to tell wp.com that the site URL from the original URL to the mapped URL (even though wp.com has the mapped domain on file). We don't really know why and causing issues on our end.

In short, our initial testing of the feature with multisite and mapped domains wasn't sufficient to expose all of these issues/questions. If so, I would have personally advocated to punt the feature until we could figure them out more.

**Comment by RavanH** [comment]
Hi @kraftbj good to hear it will not mean an indefinite postponement :) 

About other plugins that offer domain mapping: you might also need to consider (as I mentioned) multilingual plugins like Polylang and WPML. They offer an option to run different languages over different domains. Not sure if that will be detected by Jetpack on the site's own admin (remaining unaware of the different domains used on the front end) but it might cause weird results on the wp.com servers where only one front end domain is on file.

Which raises the question: Is the whole approach where a site is considered to live on only one domain not a bit limited?

**Comment by kraftbj** [comment]
Multilingual are another use case we're investigating more. So far, though, those haven't generated the support load domain mapping has. What we need to confirm is that the values we're getting on registration and are on wp.com are consistent with what are reported within the admin context.

The multilingual, from what I've seen, only modify the front-end which *should* be fine, but agreed, need to keep those in mind too.

FWIW, the approach isn't assuming a site only lives on one domain, though we are assuming a site is consistent on what it reports to us. I'm sure we'll find a solution that meets accepts that assumption.

**Comment by kraftbj** [comment]
Based on our experience with IDC UI as a whole, I'm closing this as we're going to need to rethink the approach. #3766
